libavcodec/libaribcaption.c

```diff
@@ -0,0 +1,1171 @@
+                (int)(ctx->caption.pts / 1000), (int)(ctx->caption.pts % 1000),
+                (int)((ctx->caption.wait_duration == ARIBCC_DURATION_INDEFINITE) ?
+                      -1 : ctx->caption.wait_duration / 1000),
+                (int)((ctx->caption.wait_duration == ARIBCC_DURATION_INDEFINITE) ?
+                      0 : ctx->caption.wait_duration % 1000));
+    }
+
+    switch ((enum AVSubtitleType) ctx->subtitle_type) {
+    case SUBTITLE_TEXT:
+        status = aribcaption_trans_text_subtitle(ctx);
+        break;
+
+    case SUBTITLE_ASS:
+        status = aribcaption_trans_ass_subtitle(ctx);
+        break;
+
+    case SUBTITLE_BITMAP:
+        status = aribcaption_trans_bitmap_subtitle(ctx);
+        break;
+
+    case SUBTITLE_NONE:
+    default:
+        status = 0;
+    }
+
+    if (status < 0) {
+        av_log(ctx, AV_LOG_ERROR, "Failed to set Subtitle: %s\n",
+               av_err2str(status));
+        aribcc_caption_cleanup(&ctx->caption);
+        return status;
+    }
+    if (status > 0) {
+        *got_sub_ptr = 1;
+        if (ctx->avpkt->pts != AV_NOPTS_VALUE)
+            sub->pts = av_rescale_q(ctx->avpkt->pts,
+                                    ctx->time_base, AV_TIME_BASE_Q);
+        if (ctx->caption.wait_duration == ARIBCC_DURATION_INDEFINITE)
+            sub->end_display_time = UINT32_MAX;
+        else
+            sub->end_display_time = (uint32_t)ctx->caption.wait_duration;
+    }
+
+    aribcc_caption_cleanup(&ctx->caption);
+    return avpkt->size;
+}
+
+static void aribcaption_flush(AVCodecContext *avctx)
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+
+    if (ctx->decoder)
+        aribcc_decoder_flush(ctx->decoder);
+    if (ctx->renderer)
+        aribcc_renderer_flush(ctx->renderer);
+    if (!(avctx->flags2 & AV_CODEC_FLAG2_RO_FLUSH_NOOP))
+        ctx->readorder = 0;
+}
+
+static int aribcaption_close(AVCodecContext *avctx)
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+
+    av_freep(&ctx->clut);
+    if (ctx->renderer)
+        aribcc_renderer_free(ctx->renderer);
+    if (ctx->decoder)
+        aribcc_decoder_free(ctx->decoder);
+    if (ctx->context)
+        aribcc_context_free(ctx->context);
+
+    return 0;
+}
+
+static int aribcaption_init(AVCodecContext *avctx)
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+    aribcc_profile_t profile;
+    int ret = 0;
+
+    ctx->avctx = avctx;
+
+    switch (avctx->profile) {
+    case FF_PROFILE_ARIB_PROFILE_A:
+        profile = ARIBCC_PROFILE_A;
+        /* assume 960x540 at initial state */
+        ctx->plane_width = 960;
+        ctx->plane_height = 540;
+        ctx->font_size = 36;
+        break;
+    case FF_PROFILE_ARIB_PROFILE_C:
+        profile = ARIBCC_PROFILE_C;
+        ctx->plane_width = 320;
+        ctx->plane_height = 180;
+        ctx->font_size = 16;
+        break;
+    default:
+        av_log(avctx, AV_LOG_ERROR, "Unknown or unsupported profile set.\n");
+        return AVERROR(EINVAL);
```

Comment:
Oh, you'll want to provide an AVOption to set a default profile, like in the existing decoder; default it to `a` since that's what most streams are.

Reply:
Does it mean to apply simiar patch as following?
https://lists.ffmpeg.org/pipermail/ffmpeg-devel/2022-November/303578.html

I don't undestand your intention.
On current lavf/mpegts.c code, such streams are not treated as ARIB caption.
If the profile is unkonwn, the codec id is not set, and libarib* decoder will not be called.
https://github.com/rcombs/FFmpeg/blob/5bad4856035ca5ed571e9d7d9b1d503a5c9ef0a5/libavformat/mpegts.c#L2138-L2181

Reply:
It's possible to get arib caption streams from other containers or sources (eg matroska now), or to manually instruct ffmpeg to treat a stream as containing a particular codec if it isn't detected properly (eg due to missing descriptors).

Reply:
I don't prefer such way.
Of course you can submit another patch after this patch set will be merged.

Reply:
I don't know actual code and/or status of matroska.
But I prefer such workaround will be treated in libavformat.
I think more flexible decision can be make in libavformat.